_posts/2019-03-19-til-validations-and-constraints.md

```diff
@@ -0,0 +1,229 @@
+---
+author: Prince Wilson
+date: 2019-03-19
+layout: post
+categories: til
+author_link: https://github.com/maxcell
+title:  TIL Constraints and Validations
+excerpt: >
+  Let's take a look at how Ecto handles these two ways
+  of ensuring data integrity
+---
+
+Developers want to create the best applications they can for their users.
+In the process, they want to make sure to give good feedback to their users
+when data doesn't get saved into the database. In Elixir, there is a great tool
+on top of the database that helps, Ecto! It can put validations and
+constraints onto specific fields to ensure data integrity.
+
+However, did you know they are differences between validations and constraints?
+I didn't. In the process of building a side project, I ran into the problem a few
+times! Let's talk about the goal of them and see the differences. We'll dive into
+**why we need both each** towards the end.
+
+## Data Integrity is Rule #1
+> Data Integrity is the maintenance of, and the assurance of the accuracy and
+> consistency of, data over its entire lifecycle
+> - [Wikipedia](https://en.wikipedia.org/wiki/Data_integrity)
+
+So we're building a super cool app with users that can login and logout.
+We'd probably have some schema like this:
+```elixir
+# Using Phoenix 1.4 with Contexts but still applies all the same
+defmodule MyCoolWebApp.Accounts.User do
+  use Ecto.Schema
+  import Ecto.Changeset
+
+  schema "users" do
+    field :display_name, :string
+    field :email, :string
+    field :password_hash, :string
+    field :password, :string, virtual: true
+
+    timestamps()
+  end
+end
+```
+
+And inside of it, we'd want to describe the `changeset`:
+```elixir
+def changeset(user, attrs) do
+  user
+  |> cast(attrs, [:display_name, :email, :password])
+end
+```
+
+If we only had this, we might have a few headaches come our way. It is very easy to
+prematurely submit a form without filling all the fields out. Potentially, now the
+user's profile has a lovely email or password of `nil`, or worse `""`, in the
+database. This would suck, for the user(s) and the developer(s).
+
+So in order to fix, we'd use a validation!
+
+### Validations
+Many of the validations we have in Ecto will be executed without needing
+to interact with the database. That means the validation will be executed prior to the
+attempt of inserting or updating something in the database. If we wanted to
+insert a new user into our database, we'd first want to make sure there is data
+inside of the changeset.
+
+Let's add the [`validate_require/3`](https://hexdocs.pm/ecto/Ecto.Changeset.html#validate_required/3) into our `changeset`:
+```elixir
+def changeset(user, attrs) do
+  user
+  |> cast(attrs, [:display_name, :email, :password])
+  |> validate_required([:display_name, :email, :password])
+end
+```
+
+And for free, we get a set of errors coming out if a user were to make a mistake:
```

Comment:
Instead of a "set of errors coming out", might be more clear to say that Ecto adds descriptive errors to our changeset.